File: src/distributors/mouser.rs
use crate::distributor::{Distributor, DistributorConfig, ProductInfo};
use async_trait::async_trait;
use reqwest::Client;
use serde::{Deserialize, Serialize};

// ========== Mouser API Response ==========
#[derive(Debug, Deserialize)]
pub struct MouserApiResponse {
    #[serde(rename = "Errors")]
    pub errors: Option<Vec<Error>>,

    #[serde(rename = "SearchResults")]
    pub search_results: Option<SearchResults>,
}

#[derive(Debug, Deserialize)]
pub struct Error {
    #[serde(rename = "Id")]
    pub id: i32,

    #[serde(rename = "Code")]
    pub code: String,

    #[serde(rename = "Message")]
    pub message: String,

    #[serde(rename = "ResourceKey")]
    pub resource_key: String,

    #[serde(rename = "ResourceFormatString")]
    pub resource_format_string: String,

    #[serde(rename = "ResourceFormatString2")]
    pub resource_format_string2: String,

    #[serde(rename = "PropertyName")]
    pub property_name: String,
}

#[derive(Debug, Deserialize)]
pub struct SearchResults {
    #[serde(rename = "NumberOfResult")]
    pub number_of_result: i32,

    #[serde(rename = "Parts")]
    pub parts: Vec<Part>,
}

#[derive(Debug, Deserialize)]
pub struct Part {
    #[serde(rename = "MouserPartNumber")]
    pub part_number: String,

    #[serde(rename = "ManufacturerPartNumber")]
    pub manufacturer_part_number: String,

    #[serde(rename = "Manufacturer")]
    pub manufacturer: String,

    #[serde(rename = "Description")]
    pub description: String,

    #[serde(rename = "PriceBreaks")]
    pub price_breaks: Vec<PriceBreak>,

    #[serde(rename = "AvailabilityInStock")]
    pub availability_in_stock: String,

    #[serde(rename = "unitprice")]
    pub unit_price: f64,

    #[serde(rename = "ProductDetailUrl")]
    pub product_url: String,

    #[serde(rename = "DataSheetUrl")]
    pub data_sheet_url: String,
}

#[derive(Debug, Deserialize)]
pub struct PriceBreak {
    #[serde(rename = "Quantity")]
    pub quantity: i32,

    #[serde(rename = "Price")]
    pub price: String,

    #[serde(rename = "Currency")]
    pub currency: String,
}

// Define the request body structure for the POST request.
#[derive(Serialize, Debug)]
struct MouserRequestBody {
    #[serde(rename = "SearchByPartRequest")]
    search_by_part_request: SearchByPartRequest,
}

#[derive(Serialize, Debug)]
struct SearchByPartRequest {
    #[serde(rename = "mouserPartNumbers")]
    part_numbers: Vec<String>,

    #[serde(rename = "apiKey")]
    api_key: String,
}

// ====================

pub struct MouserDistributor {
    api_key: String,
    client: Client,
    base_url: String,
}

impl MouserDistributor {
    pub fn new() -> Self {
        Self {
            api_key: String::new(),
            client: Client::new(),
            base_url: String::new(),
        }
    }
    async fn perform_query_mouser(
        &self,
        part_number: &str,
    ) -> Result<MouserApiResponse, reqwest::Error> {
        let url = format!(
            "{}/search/partnumber?apiKey={}",
            self.base_url, self.api_key
        );
        // Create the request body using the part number and API key
        let request_body = MouserRequestBody {
            search_by_part_request: SearchByPartRequest {
                part_numbers: vec![part_number.to_string()],
                api_key: self.api_key.clone(),
            },
        };
        let response = self.client.post(url).json(&request_body).send().await?;
        response.json::<MouserApiResponse>().await
    }
}

#[async_trait]
impl Distributor for MouserDistributor {
    fn initialize(&mut self, config: DistributorConfig) {
        if let DistributorConfig::ApiKey { api_key } = config {
            self.api_key = api_key;
            self.base_url = "https://api.mouser.com/api/v1/".to_string();
        }
    }

    async fn query_product_info(&self, part_number: &str) -> Option<ProductInfo> {
        match self.perform_query_mouser(part_number).await {
            Ok(response) => {
                if let Some(search_results) = response.search_results {
                    if let Some(part) = search_results.parts.get(0) {
                        return Some(ProductInfo {
                            part_number: part.part_number.clone(),
                            manafacturer_part_number: Some(part.manufacturer_part_number.clone()),
                            description: Some(part.description.clone()),
                            manafacturer: Some(part.manufacturer.clone()),
                            unit_price: Some(part.unit_price),
                            stock: Some(part.availability_in_stock.parse().unwrap_or(0)),
                            datasheet_url: Some(part.data_sheet_url.clone()),
                            product_url: Some(part.product_url.clone()),
                        });
                    }
                }
                None
            }
            Err(_) => None,
        }
    }
}

File: internal/distributors/mouser.go
package distributors

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strconv"
	"strings"
)

// Mouser represents a distributor that implements the Distributor interface
type Mouser struct {
	BaseURL string
	APIKey  string
}

// Initialize sets the API key for Mouser
func (m *Mouser) Initialize(params ...string) error {
	if len(params) < 1 {
		return fmt.Errorf("Mouser initialization requires an API key")
	}
	apiKey := params[0]

	if apiKey == "" {
		return fmt.Errorf("API key cannot be empty")
	}

	m.APIKey = apiKey
	m.BaseURL = "https://api.mouser.com/api/v1/"
	return nil
}

// QueryPartNumber queries Mouser's API for a given part number
func (m *Mouser) QueryPartNumber(partNumber string) (PartInfo, error) {
	// Handle a empty PN
	if partNumber == "" {
		return PartInfo{}, fmt.Errorf("part number cannot be empty")
	}

	url := fmt.Sprintf("%s/search/partnumber?apiKey=%s", m.BaseURL, m.APIKey)

	// Create the JSON request body
	requestBody, err := json.Marshal(map[string]interface{}{
		"SearchByPartRequest": map[string]interface{}{
			"mouserPartNumber": partNumber,
		},
	})
	if err != nil {
		return PartInfo{}, fmt.Errorf("failed to marshal JSON request body: %v", err)
	}

	// Make the POST request
	resp, err := http.Post(url, "application/json", bytes.NewBuffer(requestBody))
	if err != nil {
		return PartInfo{}, fmt.Errorf("failed to make POST request: %v", err)
	}
	defer resp.Body.Close()

	// Check if the request was successful
	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return PartInfo{}, fmt.Errorf("Mouser API request failed with status %d: %s", resp.StatusCode, string(body))
	}

	// Parse the JSON response
	var result struct {
		Errors []struct {
			Id                    int    `json: Id`
			Code                  string `json:"Code"`
			Message               string `json:Message`
			ResourceKey           string `json:ResourceKey`
			ResourceFormatString  string `json: ResourceFormatString`
			ResourceFormatString2 string `json:ResourceFormatString2`
			PropertyName          string `json:PropertyName`
		} `json:"Errors`
		SearchResults struct {
			NumberOfResult int `json:"NumberOfResult"`
			Parts          []struct {
				PartNumber             string `json:"MouserPartNumber"`
				ManufacturerPartNumber string `json:"ManufacturerPartNumber"`
				Manufacturer           string `json:"Manufacturer"`
				Description            string `json:"Description"`
				PriceBreaks            []struct {
					Quantity int    `json:"Quantity`
					Price    string `json:"Price`
					Currency string `json:"Currency`
				}
				AvailabilityInStock string  `json:"AvailabilityInStock"`
				UnitPrice           float64 `json:"unitprice"`
				ProductURL          string  `json:"ProductDetailUrl"`
				DataSheetURL        string  `json:"DataSheetUrl"`
			} `json:"Parts"`
		} `json:"SearchResults`
	}

	err = json.NewDecoder(resp.Body).Decode(&result)
	if err != nil {
		return PartInfo{}, fmt.Errorf("failed to decode JSON response: %v", err)
	}

	if len(result.Errors) != 0 {
		return PartInfo{}, fmt.Errorf("encountered a Mouser error regarding %s", result.Errors[0].PropertyName)
	}

	if result.SearchResults.NumberOfResult == 0 {
		return PartInfo{}, fmt.Errorf("no parts found for part number: %s", partNumber)
	}

	priceStr := strings.TrimPrefix(result.SearchResults.Parts[0].PriceBreaks[0].Price, "$")
	price, err := strconv.ParseFloat(priceStr, 64)
	if err != nil {
		return PartInfo{}, fmt.Errorf("Error converting unit price to float64")
	}

	availability, err := strconv.Atoi(result.SearchResults.Parts[0].AvailabilityInStock)
	if err != nil {
		return PartInfo{}, fmt.Errorf("Error converting availability to int")
	}
	// Convert to the unified format
	partInfo := PartInfo{
		PartNumber:             result.SearchResults.Parts[0].PartNumber,
		ManufacturerPartNumber: result.SearchResults.Parts[0].ManufacturerPartNumber,
		Manufacturer:           result.SearchResults.Parts[0].Manufacturer,
		Description:            result.SearchResults.Parts[0].Description,
		UnitPrice:              price,
		Availability:           availability,
		ProductURL:             result.SearchResults.Parts[0].ProductURL,
		DataSheetURL:           result.SearchResults.Parts[0].DataSheetURL,
	}

	return partInfo, nil
}
